run/websockets/app.js

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const {redisClient, getRoomFromCache, addMessageToCache} = require('./redis');
+const {addUser, getUser, deleteUser} = require('./users');
+const express = require('express');
+
+const app = express();
+app.use(express.static(__dirname + '/public'));
+
+// Serve frontend
+app.get('/', async (req, res) => {
+  res.render('index');
+});
+
+// [START cloudrun_websockets_server]
+// Initialize Socket.io
+const server = require('http').Server(app);
+const io = require('socket.io')(server);
+
+// [START cloudrun_websockets_redis_adapter]
+const redisAdapter = require('@socket.io/redis-adapter');
+// Replace in-memory adapter with Redis
+io.adapter(redisAdapter(redisClient, redisClient.duplicate()));
+// [END cloudrun_websockets_redis_adapter]
+
+// Listen for new connection
+io.on('connection', socket => {
+  // Add listener for "login" event
+  socket.on('login', async ({name, room}, callback) => {
+    try {
+      // Record socket ID to user's name and chat room
+      addUser(socket.id, name, room);
+      // Call join to subscribe the socket to a given channel
+      socket.join(room);
+      // Emit notification event
+      socket.in(room).emit('notification', {
+        title: "Someone's here",
+        description: `${name} just entered the room`,
+      });
+      // Retrieve room's message history or return null
+      const messages = await getRoomFromCache(room);
+      // Send room's message history
+      callback(null, messages);
+    } catch (err) {
+      callback(err, null);
+    }
+  });
+
+  // [START cloudrun_websockets_update_socket]
+  // Add listener for "updateSocketId" event
+  socket.on('updateSocketId', async ({name, room}) => {
+    try {
+      addUser(socket.id, name, room);
+      socket.join(room);
+    } catch (err) {
+      console.log(err);
+    }
+  });
+  // [END cloudrun_websockets_update_socket]
+
+  // Add listener for "sendMessage" event
+  socket.on('sendMessage', (message, callback) => {
```

Comment:
thought: At first I thought it was odd to see the callback model, but doing a bit of reading realize promises are not considered generally a good fit for even handling. If this sample has the goal of introducing Socket.io to developers, it may be worth adding a point about this to the comments or docs.